### PR TITLE
Storage: refactor filtering, improve performance

### PR DESCRIPTION
**Pull request by ArturWierzbicki**
Extracted performance improvement/internal refactor of storage API from auth PR https://github.com/grafana/grafana/pull/46961

### CODE REVIEW

File: pkg/infra/filestorage/filter.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+type PathFilter interface {
+	IsAllowed(path string) bool
+	ToString() string
+
+	asSQLFilter() accesscontrol.SQLFilter
+}
+
+func NewPathFilter(allowedPrefixes []string, allowedPaths []string, disallowedPrefixes []string, disallowedPaths []string) PathFilter {
+	if allowedPrefixes != nil && allowedPaths != nil && (len(allowedPrefixes)+len(allowedPaths) == 0) {
```
Comment (FZambia) [inline pkg/infra/filestorage/filter.go:27]: Can this be simplified to sth like:

```
if  len(allowedPrefixes) + len(allowedPaths) == 0 {
```

?

Or we need to distinguish between nil slice and initialized slice with zero length?

Comment (ArturWierzbicki) [reply pkg/infra/filestorage/filter.go:31]: Added a comment explaining the difference between `nil` and empty slices

**Review by FZambia — APPROVED** [review]
👍

**Review by ryantxu — APPROVED** [review]
looks great

**Pull request merged**